pkg/controllers/execution/execution_controller.go

```diff
@@ -111,20 +173,71 @@ func (c *Controller) SetupWithManager(mgr controllerruntime.Manager) error {
 		Complete(c)
 }
 
-func (c *Controller) syncWork(clusterName string, work *workv1alpha1.Work) (controllerruntime.Result, error) {
+func (c *Controller) syncWork(key util.QueueKey) error {
+	req, ok := key.(controllerruntime.Request)
+	if !ok {
+		klog.Warningf("Skip sync work for key(%+v) is not controllerruntime.Request type", key)
+		return nil
+	}
+
+	klog.Infof("Begin to sync work %s/%s", req.Namespace, req.Name)
+
 	start := time.Now()
-	err := c.syncToClusters(clusterName, work)
+
+	work := &workv1alpha1.Work{}
+	if err := c.Client.Get(c.Ctx, req.NamespacedName, work); err != nil {
+		// The resource may no longer exist, in which case we stop processing.
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+
+		return err
+	}
+
+	clusterName, err := names.GetClusterName(work.Namespace)
+	if err != nil {
+		klog.Errorf("Failed to get member cluster name for work %s/%s", work.Namespace, work.Name)
+		return err
+	}
+
+	cluster, err := util.GetCluster(c.Client, clusterName)
+	if err != nil {
+		klog.Errorf("Failed to get the given member cluster %s", clusterName)
+		return err
+	}
+
+	if !work.DeletionTimestamp.IsZero() {
+		// Abort deleting workload if cluster is unready when unjoining cluster, otherwise the unjoin process will be failed.
+		if util.IsClusterReady(&cluster.Status) {
+			err := c.tryDeleteWorkload(clusterName, work)
+			if err != nil {
+				klog.Errorf("Failed to delete work %v, namespace is %v, err is %v", work.Name, work.Namespace, err)
+				return err
+			}
+		} else if cluster.DeletionTimestamp.IsZero() { // cluster is unready, but not terminating
+			return fmt.Errorf("cluster(%s) not ready", cluster.Name)
```

Comment:
https://github.com/karmada-io/karmada/blob/1d2028b10f9b9d9ca30fe8e963b4557d347c8f51/pkg/util/worker.go#L20-L21

https://github.com/karmada-io/karmada/blob/1d2028b10f9b9d9ca30fe8e963b4557d347c8f51/pkg/util/worker.go#L122-L128


Using the `AsyncWorker` pattern seems to have a problem.
 If a cluster is in the NotReady state for up to 2 hours, and during this period, a resource is deleted from Karmada. When the cluster recovers, it seems that the member resource will not be deleted.

Reply:
Yes, there was already such a problem before this modification. I think it is necessary for the administrator to intervene manually. Otherwise, we will have to retry indefinitely because the time it takes for users to fix the cluster can vary and cannot be determined.

Reply:
In fact, we have made the parameters about backoff or retry customized in our own environment.  
Maybe we could open another issue and invite @zach593 to talk about it.

Reply:
>  there was already such a problem before this modification.

`Controller Runtime` uses `workqueue.DefaultControllerRateLimiter`, which does not have a limit on the number of times. It will perform exponential backoff.

Reply:
Can we consider modifying the implementation of `asyncWorker`?

https://github.com/karmada-io/karmada/blob/bd0c511753a9f20a474af1777695483bf3e80d27/pkg/util/worker.go#L56

Reply:
Maybe we can open a new issue to trace this problem.

Reply:
> Can we consider modifying the implementation of asyncWorker?

For me, considering whether to use `asyncWorker` is the right thing to do. It's a significant change.

/cc @RainbowMango I can't make up my mind about this issue. Can you help me take another look at it? https://github.com/karmada-io/karmada/pull/3808#discussion_r1295679640

Reply:
> Yes, there was already such a problem before this modification.

@XiShanYongYe-Chang Can you show me the code where the problem exists? 

> For me, considering whether to use asyncWorker is the right thing to do. It's a significant change.

Yeah, I agree with you, I'm still thinking if there is a better way, such as using the reconciliation mechanism provided by controller-runtime.

Reply:
When using asyncWorker, if it fails to process a certain number of times due to its maximum retry limit, the event will be discarded.

https://github.com/karmada-io/karmada/blob/bd0c511753a9f20a474af1777695483bf3e80d27/pkg/controllers/status/work_status_controller.go#L131

https://github.com/karmada-io/karmada/blob/bd0c511753a9f20a474af1777695483bf3e80d27/pkg/controllers/status/work_status_controller.go#L224

The possible reason for the update error could be that the member cluster does not allow updates within a certain period of time, which exceeds the total retry time of the worker.

Reply:
Therefore, I think that this issue can be considered separately and not addressed in this PR.

Reply:
> Therefore, I think that this issue can be considered separately and not addressed in this PR.

I'm okay to address this problem with another PR, but shall we fix it before this PR?

My concern is that this problem is being magnified, before this PR it only affected the status collection(and the impact might automatically disappear by further events), but now it also affects the `Work` synchronization.

Reply:
IMOP, the scope of impact has not expanded, but rather the problem has shifted (the synchronization of work from resourceBinding is not applicable to error scenarios such as cluster failures), and the conditions under which the problem itself occurs are relatively strict. It requires a situation where cluster failures occur during the synchronization of work triggered by changes in member cluster resources. Therefore, the scope of impact for this issue is limited.

Therefore, I think that the issue being discussed can be advanced separately. In addition, the `asyncWorker` tool is currently widely used by us and its improvement measures need to be considered as a whole.